### PR TITLE
Add alternative search providers

### LINKED ISSUE #2284: Change search provider
I know this was a feature back in the v2 days, however it would be nice to see it return in v12. No hurry, but being able to change the google search bar at the bottom to something like duckduckgo or even just local device/apps search would be nice for privacy reasons among others.
And thanks for taking the time to work on such an awesome open source project! y'all rock.

### LINKED ISSUE #2405: Alternative search providers for the search bar
Please provide an option to change the search provider to something like duckduckgo, startpage.
It would be better if the user had granular control over this like having a user defined search provider which can be used to random sites with search capacity, like Wikipedia, YouTube, etc or even a personal, public searx instance.

### LINKED ISSUE #2450: [Lawnchair 12] Reimplement alternative search engines in search bar
It'd be great if I could switch to DuckDuckGo (or any other search engine) for the search bar, rather than being stuck with Google.

Thanks!

### PR DESCRIPTION
**Pull request by evowizz**
Add alternative search providers for the hotseat search bar.

The option currently supports:
- Google
- Google Go
- DuckDuckGo

More may be added in the future (in `QsbSearchProvider`, but the goal of this PR is mainly to support the feature before adding more providers.

Fix #2284, fix #2405, fix #2450

**TODO:**

- [x]  Change logo depending on search provider
- [x]  Hide mic icon if provider doesn't support mic
- [x] _Grab some hot chocolate (that's just for me, keep reading)_
- [x] Add ability to open websites instead of apps
- [x]  Figure out how to handle the providers when the website switch is toggled and vice versa
- [x]  Add App Search provider
- [x] Export strings
- [x] Improve QsbSearchProvider

### CODE REVIEW

**Comment by evowizz** [comment]
Provider's logo are now supported. This includes theming too, however the library `com.github.devsideal:VectorChildFinder` is getting old and some parts of its code are no longer reliable as explained to @patrykmichalik.

**Comment by evowizz** [comment]
This last commit adds support for different voice intent depending on which provider is selected. If the search provider doesn't support any voice query, the mic is disabled. As suggested by @patrykmichalik, the Google Go intent uses the extra "openMic" as it doesn't have any action for launching built it assistant otherwise.

I'm open to any suggestions regarding this PR.

**Comment by evowizz** [comment]
Using IDs instead of the index allows the position of the provider to be ignored. Currently, the list is:
- Google
- Google Go
- DuckDuckGo

If the feature was released with the index instead of the IDs, it means that adding a new search provider between Google Go and DuckDuckGo would cause DuckDuckGo to have the index 3 instead of 2, so users who selected DuckDuckGo as their provider would have seen the new provider after updating instead. 

Using IDs prevents this issue from happening.

**Comment by evowizz** [comment]
I know I know, force-push is bad, but hey it's still my own branch for now

File: lawnchair/src/app/lawnchair/preferences2/PreferenceManager2.kt [hunk 1]
```diff
@@ -60,6 +62,20 @@ class PreferenceManager2(private val context: Context) : PreferenceManager {
...
+        defaultValue = QsbSearchProvider.resolveDefault(context),
+        parse = { QsbSearchProvider.fromId(it) },
+        save = { it.id },
+        onSet = { reloadHelper.recreate() }
+    )
+
+    val hotseatQsbForceWebsite = preference(
+        key = booleanPreferencesKey(name = "dock_search_bar_force_website"),
+        defaultValue = false,
+        onSet = { reloadHelper.recreate() }
```
Comment (ghost) [inline lawnchair/src/app/lawnchair/preferences2/PreferenceManager2.kt]:
```suggestion
```

File: lawnchair/src/app/lawnchair/qsb/QsbLayout.kt [hunk 2]
```diff
@@ -54,46 +51,36 @@ class QsbLayout(context: Context, attrs: AttributeSet?) : FrameLayout(context, a
...
-        } else {
-            micIcon.setIcon(isGoogle = false, themed = false)
-            with(gIcon) {
-                setImageResource(R.drawable.ic_qsb_search)
-                setColorFilter(Themes.getColorAccent(context))
-            }
-        }
-    }
+        val supportsLens = searchProvider == Google
+
```
Comment (ghost) [inline lawnchair/src/app/lawnchair/qsb/QsbLayout.kt:61]:
```suggestion

        preferenceManager2.hotseatQsbForceWebsite.onEach(launchIn = viewAttachedScope) { hotseatQsbForceWebsite ->
            setUpMainSearch(
                searchProvider = searchProvider,
                forceWebsite = hotseatQsbForceWebsite,
            )
        }

```

File: lawnchair/src/app/lawnchair/qsb/QsbLayout.kt [hunk 3]
```diff
@@ -54,46 +51,36 @@ class QsbLayout(context: Context, attrs: AttributeSet?) : FrameLayout(context, a
         preferenceManager = PreferenceManager.getInstance(context)
         preferenceManager2 = PreferenceManager2.getInstance(context)
 
-        val searchPackage = getSearchPackageName(context)
-        setUpMainSearch(searchPackage)
+        val searchProvider = getSearchProvider(context, preferenceManager2)
+        val forceWebsite = preferenceManager2.hotseatQsbForceWebsite.firstBlocking()
+        setUpMainSearch(searchProvider, forceWebsite)
```
Comment (ghost) [inline lawnchair/src/app/lawnchair/qsb/QsbLayout.kt]:
```suggestion
```

**Review by ghost — CHANGES_REQUESTED** [review]
For better performance, instead of calling `ReloadHelper#recreate` when the `hotseatQsbForceWebsite` setting is changed, I’d suggest using `Preference#onEach`.

**Comment by evowizz** [comment]
No problem, done 👍

File: lawnchair/res/values/strings.xml [hunk 4]
```diff
@@ -210,4 +210,10 @@
     <string name="icon_picker_description">May cause Lawnchair to freeze.</string>
     <string name="font_picker_label">Font Customization</string>
     <string name="font_picker_description">Some text remains unchanged.</string>
+    <string name="search_provider_app_search">App Search</string>
+    <string name="search_provider_google" translatable="false">Google</string>
+    <string name="search_provider_google_go" translatable="false">Google Go</string>
+    <string name="search_provider_duckduckgo" translatable="false">DuckDuckGo</string>
+    <string name="always_open_website">Always Open Website</string>
```
Comment (ghost) [inline lawnchair/res/values/strings.xml]:
```suggestion
    <string name="always_open_website_label">Always Open Website</string>
    <string name="always_open_website_description">Open search provider’s website even if their app is installed.</string>
```

File: lawnchair/src/app/lawnchair/qsb/providers/AppSearch.kt [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package app.lawnchair.qsb.providers
+
+import app.lawnchair.qsb.ThemingMethod
+import com.android.launcher3.R
+
+object AppSearch : QsbSearchProvider(
+    id = "app-search",
```
Comment (ghost) [inline lawnchair/src/app/lawnchair/qsb/providers/AppSearch.kt]:
```suggestion
    id = "app_search",
```

File: lawnchair/src/app/lawnchair/ui/preferences/DockPreferences.kt [hunk 6]
```diff
@@ -92,6 +102,15 @@ fun DockPreferences() {
                             valueRange = 0F..1F,
                             showAsPercentage = true,
                         )
+                        QsbProviderPreference(
+                            value = hotseatQsbProvider,
+                            edit = { hotseatQsbProvider.set(value = it) },
+                        )
+                        SwitchPreference2(
+                            checked = hotseatQsbUseWebsite,
+                            label = stringResource(R.string.always_open_website),
```
Comment (ghost) [inline lawnchair/src/app/lawnchair/ui/preferences/DockPreferences.kt]:
```suggestion
                            label = stringResource(R.string.always_open_website_label),
                            description = stringResource(R.string.always_open_website_description),
```

**Review by ghost — CHANGES_REQUESTED** [review]
Two small suggestions:
- I’d suggest adding a description for the Always Open Website setting in order to make it clearer what the switch does.
- I’d suggest changing the ID of `AppSearch` from “app-search” to “app_search” so that it is consistent with the ID of `GoogleGo`.

File: lawnchair/src/app/lawnchair/qsb/providers/AppSearch.kt (hunk 5 above)
Comment (evowizz) [reply lawnchair/src/app/lawnchair/qsb/providers/AppSearch.kt]: Well the ID isn't that important, but for consistency, sure.

**Review by ghost — APPROVED** [review]

**Comment by ghost** [comment]
Thanks a bunch, @evowizz!

**Comment by robotwombat** [comment]
Thanks for all of the work!

One thing though. As of right now, with my provider set to DuckDuckGo, tapping on the search bar just takes me to the website. I don't have the ability to enter any text and then submit. I'm only able to tap on the search bar, and it then takes me to duckduckgo.com, then I have to tap on the search bar within the website to search for something.

Am I doing something wrong? Thanks

EDIT:
So installing the DDG app makes it work, but I'd rather not use their browser. Is there a way to make it work correctly with Firefox? (Or whatever your default browser is)

Thanks again

**Comment by RikKoedoot** [comment]
In Lawnchair V10 we did have the possibility of searching DDG via the browser. Maybe that behavior could be ported.

**Comment by ljgp6** [comment]
Please also include petal search as a provider and an option to change the provider for mic(google) from assistant to google voice search. thank you

**Comment by evowizz** [comment]
This Pull Request has now been merged. You all may want to create an issue for all of this instead of commenting here.

**Pull request merged**